ldap/client/api/src/main/java/org/apache/directory/ldap/client/api/LdapNetworkConnection.java

```diff
@@ -599,102 +604,100 @@ public boolean connect() throws LdapException
 
         // We try to initialize a socket connection with all the IPs simultaneously, picking the first that succeeds and dropping the others.
         ConnectFuture connectionFuture = null;
-        LdapNetworkConnectionFuture connFuture = null;
         final long giveUpAfter = System.currentTimeMillis() + timeout * 2;
-        ExecutorService executor = Executors.newFixedThreadPool( addresses.length + 1 ); // Just in case of overlapping.
-        while ( connFuture == null && giveUpAfter > System.currentTimeMillis() )
+        this.completionService = new ExecutorCompletionService<>( executor );
+        Set<Future<LdapNetworkConnectionFuture>> futures = new HashSet<>();
+        for ( int i = 0; i < addresses.length; i++ )
         {
-            LinkedList<LdapNetworkConnectionFuture> connFutures = new LinkedList<>();
-            for ( int i = 0; i < addresses.length; i++ )
-            {
-                // The IP to use during this iteration.
-                final InetAddress addr = addresses[i];
+            // The IP to use during this iteration.
+            final InetAddress addr = addresses[i];
 
-                // Build the connection address
-                final SocketAddress address = new InetSocketAddress( addr, config.getLdapPort() );
+            // Build the connection address
+            final SocketAddress address = new InetSocketAddress( addr, config.getLdapPort() );
 
-                try
-                {
-                    LdapNetworkConnectionFuture cf = connectFuture( address );
-                    executor.execute( cf.getConnectFuture() );
-                    connFutures.add( cf );
-                }
-                catch ( InterruptedException e )
-                {
-                    LOG.info( "InterruptedException occurred when connecting to " + address + " - Message: " + e.getMessage() );
-                }
+            try
+            {
+                futures.add( connectFuture( address ) );
             }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when connecting to " + address + " - Message: " + e.getMessage() );
+            }
+        }
 
-            // Wait for the futures to complete. Break out early if there is a successful connection.
-            breakaway:
-            while ( connFutures.size() > 0 )
+        // Wait for the futures to complete. Break out early if there is a successful connection.
+        while ( futures.size() > 0 && giveUpAfter > System.currentTimeMillis() )
+        {
+            Future<LdapNetworkConnectionFuture> completedFuture;
+            try
             {
-                for ( LdapNetworkConnectionFuture cf : connFutures )
-                {
-                    if ( !cf.getConnectFuture().isDone() )
-                    {
-                        continue;
-                    }
+                completedFuture = completionService.take();
```

Comment:
doesn't take block forever if it takes forever? I think you can use `poll(timeout)` instead